paasta_tools/autoscaling_lib.py

```diff
@@ -508,41 +512,130 @@ def spotfleet_scaler(resource, error):
 
     if new_capacity != current_capacity:
         print "Scaling SFR %s from %d to %d!" % (resource['id'], current_capacity, new_capacity)
-        scale_aws_spot_fleet_request(sfr_id=resource['id'], target_capacity=new_capacity, ec2_client=ec2_client)
+        return new_capacity - current_capacity, new_capacity
+    else:
+        return 0
+
+
+def slave_pid_to_ip(slave_pid):
+    """Convert slave_pid to IP
+
+    :param: slave pid e.g. slave(1)@10.40.31.172:5051
+    :returns: ip address"""
+    regex = re.compile(r'.+?@([\d\.]+):\d+')
+    return regex.match(slave_pid).group(1)
+
+
+def pick_slaves_to_kill(mesos_state, number, pool='default'):
+    """Pick the best slaves to kill. This returns a given number of slaves
+    after sorting by number of mesos tasks running. It sorts first by number
+    of chronos tasks, then by total number of tasks
+
+    :param mesos_state: mesos_state dict
+    :param number: number of slaves to return
+    :param pool: pool of slaves to consider
+    :returns: list of slave_pids"""
+    slaves = get_mesos_task_count_by_slave(mesos_state, pool=pool)
+    if slaves:
+        slaves_by_task_count = [slave['slave'] for slave in sorted(slaves.values(),
+                                                                   key=lambda x: (x['chronos_count'], x['count']))]
+        slaves_to_kill = [slave['pid'] for slave in slaves_by_task_count]
+        return slaves_to_kill[0:number]
     else:
-        print "No need to scale. new_capacity (%d) matches current capacity (%d)" % (new_capacity, current_capacity)
+        return []
 
 
-def scale_aws_spot_fleet_request(sfr_id, target_capacity, ec2_client):
-    return ec2_client.modify_spot_fleet_request(SpotFleetRequestId=sfr_id, TargetCapacity=target_capacity)
-    #########
-    # TODO: determine which boxes to kill first
-    # TODO: drain boxes/maintainance mode before killing
-    #########
-    # http://boto3.readthedocs.org/en/latest/reference/services/ec2.html#EC2.Client.modify_spot_fleet_request
-    # describes how to use boto to scale down spot fleet requests without
-    # killing instance using the noTermination
-    # ExcessCapacityTerminationPolicy
-    #########
+def kill_instances_gracefully(instances_to_kill):
+    """Kills instances after waiting for slave to be running 0 mesos tasks
+
+    :param instances_to_kill: dict of instances to kill {<slave_id>: <instance_id>}"""
+    ec2_client = boto3.client('ec2')
+    try:
+        with Timeout(seconds=CLUSTER_DRAIN_TIMEOUT):
+            while instances_to_kill:
+                mesos_state = get_mesos_state_from_leader()
+                slaves_count = get_mesos_task_count_by_slave(mesos_state, pool=None)
+                for slave_id, instance_id in instances_to_kill.items():
+                    num_tasks = slaves_count[slave_id]['count']
+                    log.debug("Instance: {0}, still running {1} tasks".format(instance_id, num_tasks))
```

Comment:
drop the 'still' here, and can you add a debug line for when you terminate please